src/Core/Managed/Shared/Extensibility/MetricManager.cs

```diff
@@ -0,0 +1,285 @@
+            }
+
+            return this.metricDictionary.GetOrAdd(metric, (m) => { return new SimpleMetricStatisticsAggregator(); });
+        }
+
+        /// <summary>
+        /// Calculates wait time until next snapshot of the aggregators.
+        /// </summary>
+        /// <returns>Wait time.</returns>
+        private static TimeSpan GetWaitTime()
+        {
+            DateTimeOffset currentTime = DateTimeOffset.UtcNow;
+
+            double minutesFromZero = currentTime.Subtract(DateTimeOffset.MinValue).TotalMinutes;
+
+            // we want to wake up exactly at 1 second past minute
+            // to make perceived system latency look smaller
+            var nextWakeTime = DateTimeOffset.MinValue
+                .AddMinutes((long)minutesFromZero)
+                .Add(snapshotFrequency)
+                .AddSeconds(1);
+
+            TimeSpan sleepTime = nextWakeTime - DateTimeOffset.UtcNow;
+
+            // adjust wait time to a bit longer than a minute if the wake up time is within few seconds from now
+            return sleepTime < TimeSpan.FromSeconds(3) ? sleepTime.Add(snapshotFrequency) : sleepTime;
+        }
+
+        /// <summary>
+        /// Generates telemetry object based on the metric aggregator.
+        /// </summary>
+        /// <param name="metric">Metric definition.</param>
+        /// <param name="statistics">Metric aggregator statistics calculated for a period of time.</param>
+        /// <returns>Metric telemetry object resulting from aggregation.</returns>
+        private static AggregatedMetricTelemetry CreateAggergatedMetricTelemetry(Metric metric, SimpleMetricStatisticsAggregator statistics)
+        {
+            var telemetry = new AggregatedMetricTelemetry(
+                metric.Name,
+                statistics.Count,
+                statistics.Sum,
+                statistics.Min,
+                statistics.Max,
+                statistics.StandardDeviation);
+
+            if (metric.Dimensions != null)
+            {
+                foreach (KeyValuePair<string, string> property in metric.Dimensions)
+                {
+                    telemetry.Properties.Add(property);
+                }
+            }
+
+            return telemetry;
+        }
+
+        private void Flush(bool disposing)
+        {
+            if ((!disposing) && this.disposed)
+            {
+                throw new ObjectDisposedException(nameof(MetricManager));
+            }
+
+            try
+            {
+                this.Snapshot();
+                this.telemetryClient.Flush();
+            }
+            catch (Exception ex)
+            {
+                CoreEventSource.Log.FailedToFlushMetricAggregators(ex.ToString());
+            }
+        }
+
+        /// <summary>
+        /// Takes a snapshot of aggregators collected by this instance of the manager
+        /// and schedules the next snapshot.
+        /// </summary>
+        private Task SnapshotAndReschedule()
```

Comment:
Having examined the DelayTask (I am not sure whether I agree with Oleg about using it, but he makes some good points), I think that running is underStartNew is no required. DelayTask calls Task.Delay and then runs the delegate synchronously in respect to the completion of the Delay. Since we are not scheduling the timer again until we finsih the snapshot, there is no reason for us to just threads here.

I would just do

```
return () =>
    {
        // all the shapshot stuff
        return Task.FromResult(1);
    };
```